src/engagement_db_to_analysis/engagement_db_to_analysis.py

```diff
@@ -135,28 +135,32 @@ def _fold_messages_by_uid(user, messages_traced_data):
         participant_uuid = message["participant_uuid"]
         message_dataset = message["dataset"]
 
-        if message["participant_uuid"] not in participants_traced_data_map.keys():
+        # Create an empty TracedData for this participant if this participant hasn't been seen yet.
+        if participant_uuid not in participants_traced_data_map.keys():
+            participants_traced_data_map[participant_uuid] = \
+                TracedData({}, Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string()))
+
+        # Get the existing list of messages for this dataset, if it exists, otherwise initialise with []
+        participant_td = participants_traced_data_map[participant_uuid]
+        dataset_messages = participant_td.get(message_dataset, [])
+
+        # Append this message to the list of messages for this dataset, and write-back to TracedData.
+        dataset_messages = dataset_messages.copy()
+        dataset_messages.append(dict(message))
+        participant_td.append_data(
+            {message_dataset: dataset_messages},
+            Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string())
+        )
+        # TODO: Consider not converting messages to TracedData before this function, as that would remove the need
```

Comment:
unfortunately, we need to if we want to keep history for message filters.

Reply:
converted to an explanation